javascript/ql/src/semmle/javascript/DOM.qll

```diff
@@ -374,10 +374,26 @@ module DOM {
         this = DOM::domValueRef().getAPropertyRead("baseUri")
         or
         this = DataFlow::globalVarRef("location")
+        or
+        this = any(DataFlow::Node n | n.hasUnderlyingType("Location")).getALocalSource() and
+        not this = nonFirstLocationType(DataFlow::TypeTracker::end()) // only start from the source, and not the locations we can type-track to.
       }
     }
   }
 
+  /**
+   * Get a reference to a node of type `Location` that has gone through at least 1 type-tracking step.
+   */
+  private DataFlow::SourceNode nonFirstLocationType(DataFlow::TypeTracker t) {
+    // One step inlined in the beginning.
+    exists(DataFlow::TypeTracker t2 |
+      result =
+        any(DataFlow::Node n | n.hasUnderlyingType("Location")).getALocalSource().track(t2, t)
+    )
+    or
+    exists(DataFlow::TypeTracker t2 | result = nonFirstLocationType(t2).track(t2, t))
```

Comment:
Do we actually need the recursive case here?

If one of the type-tracking predecessors of `x` has type `Location`, then `x` is not a "first" source of the `Location` object. Do we need to check more than that?

Reply:
Great point!   

I don't think we need more than the first. I'll test it.

Reply:
No, we do need the recursive case. 

Otherwise we don't flag the access to `container.loc` from the testcase as being "redundant".   
I think it is because we need both a store-step and a load-step for the type-tracker to go through the `container` object.